core/arch/arm/plat-hikey/spi_test.c

```diff
@@ -28,73 +28,271 @@
 #include <drivers/pl022_spi.h>
 #include <drivers/pl061_gpio.h>
 #include <hikey_peripherals.h>
+#include <io.h>
+#include <kernel/tee_time.h>
 #include <stdint.h>
 #include <trace.h>
+#include <util.h>
 
-/*
- * spi_init() must be run before calling this function.
- *
- * This runs a loopback test by default, so the SPI module will just
- * receive what is transmitted, i.e. 0x01, 0x80, 0x00.
- *
- * In non-loopback mode, the transmitted value will elicit a readback of
- * the measured value from the ADC chip on the Linksprite 96Boards
- * Mezzanine card [1], which can be connected to either a sliding
- * rheostat [2] or photoresistor [3].
- *
- * [1] http://linksprite.com/wiki/index.php5?title=Linker_Mezzanine_card_for_96board
- * [2] http://learn.linksprite.com/96-board/sliding-rheostat
- * [3] http://learn.linksprite.com/96-board/photoresistor
- */
-void spi_test(void)
+#define PL022_STAT	0x00C
+#define PL022_STAT_BSY	SHIFT_U32(1, 4)
+
+static void spi_cs_callback(enum gpio_level value)
 {
-	struct pl061_data platform_pl061_data;
-	struct pl022_data platform_pl022_data;
+	static bool inited;
+	struct pl061_data pd;
 	vaddr_t gpio6_base = nsec_periph_base(GPIO6_BASE);
 	vaddr_t spi_base = nsec_periph_base(SPI_BASE);
+
+	if (!inited) {
+		pl061_init(&pd);
+		pl061_register(gpio6_base, 6);
+		pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
+		pd.chip.ops->set_interrupt(GPIO6_2, GPIO_INTERRUPT_DISABLE);
+		pd.chip.ops->set_direction(GPIO6_2, GPIO_DIR_OUT);
+		inited = true;
+	}
+
+	if (read8(spi_base + PL022_STAT) & PL022_STAT_BSY)
+		DMSG("pl022 busy - do NOT set CS!");
+	while (read8(spi_base + PL022_STAT) & PL022_STAT_BSY)
+		;
+	DMSG("pl022 done - set CS!\n");
+
+	pd.chip.ops->set_value(GPIO6_2, value);
+}
+
+static void spi_set_cs_mux(uint32_t val)
+{
+	uint32_t data;
+	vaddr_t pmx0_base = nsec_periph_base(PMX0_BASE);
+
+	if (val == PINMUX_SPI) {
+		DMSG("configure gpio6 pin2 as SPI\n");
+		write32(PINMUX_SPI, pmx0_base + PMX0_IOMG106);
+	} else {
+		DMSG("configure gpio6 pin2 as GPIO\n");
+		write32(PINMUX_GPIO, pmx0_base + PMX0_IOMG106);
+	}
+
+	data = read32(pmx0_base + PMX0_IOMG106);
+	if (data)
+		DMSG("gpio6 pin2 is SPI\n");
+	else
+		DMSG("gpio6 pin2 is GPIO\n");
+}
+
+static void spi_test_with_manual_cs_control(void)
+{
+	struct pl022_data pd;
+	vaddr_t spi_base = nsec_periph_base(SPI_BASE);
 	uint8_t tx[3] = {0x01, 0x80, 0x00};
 	uint8_t rx[3] = {0};
-	size_t i, j, num_rxpkts, len = 3;
+	size_t i, j, len = 3;
+	enum spi_result res;
+
+	spi_set_cs_mux(PINMUX_GPIO);
+
+	DMSG("set CS callback\n");
+	pd.cs_data.cs_cb = NULL;
+	pd.b_cs_callback = true;
+
+	DMSG("spi_base: 0x%" PRIxVA "\n", spi_base);
+	DMSG("configure SPI\n");
+	pd.base = spi_base;
+	pd.clk_hz = SPI_CLK_HZ;
+	pd.speed_hz = SPI_10_KHZ;
+	pd.mode = SPI_MODE0;
+	pd.data_size_bits = 8;
+	pd.loopback = true;
+
+	pl022_init(&pd);
```

Comment:
ok if you are sure this will be called only once, as it is currently the case.
maybe those need a `xxx_deinit()` function. it is likely that some late code will init again the pl022 instance. I leave it up to you to decide.